src/backend/commands/tablecmds.c

```diff
@@ -974,7 +974,8 @@ DefineRelation(CreateStmt *stmt, char relkind, Oid ownerId, char relstorage, boo
 		cooked_constraints = list_concat(cooked_constraints, newCookedDefaults);
 	}
 
-	if (stmt->attr_encodings)
+	/* Encoding clause is useless for foreign tables */
+	if (stmt->attr_encodings && !RelationIsForeign(rel))
```

Comment:
What about other kinds of relations (heap tables, external tables)? How are they handled?

Reply:
There is not any problem with other kind of relations. Extra column options do not apply if table is not `appendonly=true` or not `orientation=column`.
```
test=# create table test3 (a int) with (appendoptimized=false);
NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 'a' as the Greenplum Database data distribution key for this table.
HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
CREATE TABLE
test=# \q
...
CREATE TABLE public.test3 (
    a integer
)
WITH (appendonly='false')
 DISTRIBUTED BY (a);
```
```
test=# CREATE EXTERNAL TABLE ext_expenses (name text, date date, 
amount float4, category text, description text) 
LOCATION ( 
'file://seghost1/dbfast/external/expenses1.csv'
)
FORMAT 'CSV' ( HEADER );
NOTICE:  HEADER means that each one of the data files has a header row
CREATE EXTERNAL TABLE
...
CREATE EXTERNAL TABLE public.ext_expenses (
    name text,
    date date,
    amount real,
    category text,
    description text
) LOCATION (
    'file://seghost1/dbfast/external/expenses1.csv'
) ON ALL 
FORMAT 'csv' (delimiter E',' null E'' escape E'"' quote E'"' header)
ENCODING 'UTF8';
```

Reply:
`transformAttributeEncoding()` applies `ENCODING` which is set at table columns, but not filters if it is set at `gp_default_storage_options`.

Reply:
`transformAttributeEncoding()` covers all cases. Default options are placed inside `stenc` argument.
I did a mistake in comment about `can_enc` assignment, the correct one is:
```
bool can_enc = is_aocs(stmt->options) && !cxt->isforeign;
```